src/Compiler.Dynamic/SystemWebCompilation.cs

```diff
@@ -79,7 +82,7 @@ protected IDisposable MarkRecompile()
         });
     }
 
-    public IFileProvider Files => _env.ContentRootFileProvider;
+    public IFileProvider Files => _webFormsOptions.Value.WebFormsFileProvider ?? _env.ContentRootFileProvider;
```

Comment:
no need to keep the env as well - let's have the WebFormsFileProvider be the source of truth (that'll default to environment one)